[RDY] Fix allowing arguments as part of shell for `:term` command

FriedSock opened the pull request:
Fixes the default behaviour of `:term`, especially when arguments have been set as part of the `shell` setting. Fixes https://github.com/neovim/neovim/issues/3999

FriedSock committed:
Tokenize p_sh if used as default in ex_terminal

Previously p_sh was being used as the first
argument in a list when calling termopen(), this
would try to call an untokenized version of shell,
meaning if you had an argument passed to your
shell like so:

`set shell=/bin/bash\ --login`

The command will fail as “ --login” is interpreted
as part of a file path.

FriedSock commented:
The `default` check seems to be showing for a previous commit that I force pushed over, is there a way to re-run it?

ZyX-I commented:
This is not a PR that fixes formatting, so first commit should not be here. You should not do changes to irrelevant functions because this fails lint tests (change to some line makes linter stop ignoring errors in two adjacent lines and in that line itself), makes review harder and specifically this change does not really make code look cleaner.

FriedSock commented:
sorry @ZyX-I, I have removed it now.

FriedSock committed:
Add test for `terminal` with a shell with args

FriedSock commented:
Quite confused about the build intermittently failing on gcc-5. It seems there is a segfault when trying to run the `with no argument, acts like termopen()` test.

justinmk commented on test/functional/fixtures/shell-test.c:
```diff
@@ -30,6 +32,15 @@ int main(int argc, char **argv)
   if (argc >= 2) {
+    if (strcmp(argv[1], "-t") == 0) {
+      if (argc < 3) {
+        fprintf(stderr,"Missing promt text for -t option\n");
```
typo

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -68,6 +68,18 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
+  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC
+{
+  size_t argc = tokenize((char_u*)cmd, NULL);
+  char **rv = xmalloc((argc + 4) * sizeof(*rv));
```
Is it legal to dereference `**rv` while it's being assigned?

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -68,6 +68,18 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
 }
 
+char **build_argv(char *cmd)
+  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC
```
can also add FUNC_ATTR_NONNULL_ALL here

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -68,6 +68,18 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
+
+  tokenize((char_u*)cmd, rv);
+  assert(rv[0]);
+  rv[argc] = NULL;
```
why is `cmd` being `tokenize`d twice (first time on line 74)?

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -68,6 +68,18 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
   return rv;
 }
 
+char **build_argv(char *cmd)
```
This function is nearly identical to `shell_build_argv`, why not use `shell_build_argv(cmd, NULL)` instead?

FriedSock replied:
Because that would call a shell inside a shell, which seemed wasteful, and confusing behaviour.

justinmk replied:
Those examples would be great to have in a `///` comment above each respective function.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9454,22 +9454,35 @@ static void ex_folddo(exarg_T *eap)
-  if (name != (char *)p_sh) {
-    xfree(name);
+    while (*p != NULL) {
+      snprintf(shell_argv, sizeof(shell_argv), "%s, \"%s\"", shell_argv, *p);
```
This seems like a weird way to concat a string. `vim_strcat` and `concat_str` are alternatives.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9454,22 +9454,35 @@ static void ex_folddo(exarg_T *eap)
+
+    snprintf(ex_cmd, sizeof(ex_cmd),
+             ":enew%s | call termopen([%s]) | startinsert",
+             eap->forceit == true ? "!" : "", shell_argv);
```
no need to compare booleans explicitly.

FriedSock replied:
Ha, I only changed that line because the linter complained, good spot.

justinmk commented:
>  It seems there is a segfault when trying to run the with no argument, acts like termopen() test.

Can you paste the backtrace? Or what leads you to believe it's a segfault? I only saw hung builds (maybe caused by a "Press ENTER" prompt).

FriedSock commented:
I only say that because the hung builds I was getting on clang in a previous iteration on the same test were due to a segfault. I need to get round to setting up a VM with gcc 5 (Im on os x which seems unable to build neovim on gcc 5) to run them locally to figure out the problem.

FriedSock committed:
Fix linting errors

FriedSock committed:
Remove superfluous boolean comparisons

justinmk commented on test/functional/terminal/ex_terminal_spec.lua:
```diff
@@ -42,35 +42,53 @@ describe(':terminal', function()
     screen:attach({rgb=false})
     -- shell-test.c is a fake shell that prints its arguments and exits.
-    nvim('set_option', 'shell', nvim_dir..'/shell-test')
+    nvim('set_option', 'shell', nvim_dir..'/shell-test -t jeff')
```
Can just do this at the start of a new `it()` case, instead of duplicating the full `describe()` block.

justinmk commented on test/functional/fixtures/shell-test.c:
```diff
@@ -30,7 +32,14 @@ int main(int argc, char **argv)
-    if (strcmp(argv[1], "EXE") == 0) {
+    if (strcmp(argv[1], "-t") == 0) {
+      if (argc < 3) {
+        fprintf(stderr,"Missing promt text for -t option\n");
```
typo

FriedSock committed:
Fix typo

FriedSock committed:
Expand and refactor ex terminal tests

Adds a new test that covers the case of calling a command on a shell
that has arguments.

Extract a `terminal_with_fake_shell_with_arguments` function to allow
keeping the two test cases within the original describe block.

FriedSock committed:
Add comments to build_argv and shell_build_argv

There is a subtle difference between how these two methods work. Adding
comments based on PR discussions to clarify this.

FriedSock commented:
@justinmk I have made all the changes you suggested now. Thank you for your time and patience 🙇

FriedSock committed:
Convert vim_strcat -> xstrlcat

Since the previous build, vim_strcat was deprecated, convert to use
xstrlcat

justinmk commented on test/functional/terminal/ex_terminal_spec.lua:
```diff
@@ -68,6 +68,11 @@ describe(':terminal (with fake shell)', function()
   end
 
+  local function terminal_with_fake_shell_with_arguments(cmd)
+    nvim('set_option', 'shell', nvim_dir..'/shell-test -t jeff')
```
would prefer to put this explicitly in the tests, then call the normal `terminal_with_fake_shell()`.

FriedSock replied:
👍

justinmk commented on test/functional/fixtures/shell-test.c:
```diff
@@ -12,8 +12,12 @@ static void help(void)
   puts("  shell-test EXE \"prog args...\"");
   puts("    Prints \"ready $ prog args...\\n\" to stderr.");
+  puts("  shell-test -t {prompt text} EXE \"prog args...\"");
+  puts("    Prints \"ready {prompt text} $ progs args...\" to stderr.");
```
this isn't used, is it needed?

FriedSock replied:
Ah it is used, I just made a typo. It should say:

` Prints \"{prompt text} $ progs args...\" to stderr."`

Will fix now, thanks!

justinmk commented on test/functional/terminal/ex_terminal_spec.lua:
```diff
@@ -53,6 +53,12 @@ end)
+  -- the {cmd} and exits immediately .
+  local function terminal_with_fake_shell(cmd)
+    execute("terminal "..(cmd and cmd or ""))
+  end
```
why move this?

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9624,21 +9624,41 @@ static void ex_folddo(exarg_T *eap)
+
+    if (*p != NULL) {
+      snprintf(tempstring, sizeof(tempstring), "\"%s\"", *p);
+      xstrlcat((char *const)shell_argv, (char *const)tempstring, 512);
```
change the declarations instead of casting

justinmk replied:
use `sizeof(shell_argv)` instead of hardcoding 512

FriedSock replied:
👍, Ill have to keep tempstring as a `char *` and cast it, as that is what `snprintf` requires it seems.

FriedSock replied:
oh no, I was wrong, changing the declaration as you suggested below removed the warning.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9624,21 +9624,41 @@ static void ex_folddo(exarg_T *eap)
+    char **argv = build_argv((char *)p_sh);
+    char **p = argv;
+    char tempstring[512] = "";
+    char shell_argv[512] = "";
```
this is an unusual way to initialize a char array, normally we do:

    char foo[512];

If you really want zero it out, do:

    char foo[512] = { 0 };

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9624,21 +9624,41 @@ static void ex_folddo(exarg_T *eap)
 
 static void ex_terminal(exarg_T *eap)
 {
+  // We will call termopen() with shell as an arglist if not given a {cmd}.
```
let's put this comment near the if-block

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9624,21 +9624,41 @@ static void ex_folddo(exarg_T *eap)
     mustfree = true;
-    lquote = rquote = "\"";
+    snprintf(ex_cmd, sizeof(ex_cmd),
+             ":enew%s | call termopen(\"%s\") | startinsert",
```
use  `termopen('%s')`

FriedSock replied:
Turns out that these need to be double quotes, there is a test that `:term` allows quotes and slashes that fails if I change this.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9625,20 +9625,40 @@ static void ex_folddo(exarg_T *eap)
 static void ex_terminal(exarg_T *eap)
 {
   char *name = (char *)p_sh;  // Default to 'shell' if {cmd} is not given.
+  char ex_cmd[1024] = "";
```
`= { 0 };`

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9625,20 +9625,40 @@ static void ex_folddo(exarg_T *eap)
+    }
+
+    while (*p != NULL) {
+      snprintf(tempstring, sizeof(tempstring), ", \"%s\"", *p);
```
Do we need the space after the comma?

FriedSock replied:
ah, we don't, good spot 👍

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9625,20 +9625,40 @@ static void ex_folddo(exarg_T *eap)
+      snprintf(tempstring, sizeof(tempstring), ", \"%s\"", *p);
+      xstrlcat(shell_argv, tempstring, sizeof(shell_argv));
+      p++;
+    }
```
We could remove the previous `if` block, if we use `shell_argv+2` in the following `snprintf()`.

FriedSock replied:
Wow, yeah that works, would never have thought of that, thanks 😄

justinmk commented on test/functional/terminal/ex_terminal_spec.lua:
```diff
@@ -79,6 +79,18 @@ describe(':terminal (with fake shell)', function()
     ]])
   end)
 
+  it('with no argument, but the shell has arguments, acts like termopen()', function()
```
vim options are indicated with single quotes, by convention, so this can be a bit more terse:

    it("with no argument, but 'shell' has arguments, acts like termopen()", function()

justinmk commented on test/functional/terminal/ex_terminal_spec.lua:
```diff
@@ -90,6 +102,18 @@ describe(':terminal (with fake shell)', function()
     ]])
   end)
 
+  it('executes a given command through the shell, when the shell has arguments', function()
```
it("executes a given command through the shell, when 'shell' has arguments', function()

justinmk commented:
LGTM after remaining comments.

FriedSock committed:
Optimize `ex_terminal`

We can remove an if statement by shifting the pointer for the shell_arg
string past the leading comma. We also do not need spaces in shellargv.

We can also remove some array initialisation and a branch for some late
memory freeing.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9625,20 +9625,40 @@ static void ex_folddo(exarg_T *eap)
+             ":enew%s | call termopen(\"%s\") | startinsert",
+             eap->forceit ? "!" : "", name);
+  } else {
+    char **argv = build_argv((char *)p_sh);
```
What happens if `p_sh` is an empty string?

FriedSock replied:
It looks like you enter a command you will get a message: `execvp failed: No such file or directory: ''`. I think people should really expect an error if they set `shell=''` and try to run the terminal.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9625,20 +9625,34 @@ static void ex_folddo(exarg_T *eap)
 static void ex_terminal(exarg_T *eap)
 {
   char *name = (char *)p_sh;  // Default to 'shell' if {cmd} is not given.
```
`name` and the `xfree(name);` at the end can be moved into the first `if` branch.

FriedSock replied:
I did look at this in a previous iteration, but there were some test failures. It may have been due to something else, ill have another look.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9625,20 +9625,34 @@ static void ex_folddo(exarg_T *eap)
 {
   char *name = (char *)p_sh;  // Default to 'shell' if {cmd} is not given.
+  char ex_cmd[1024] = { 0 };
   bool mustfree = false;
```
Can be removed.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9625,20 +9625,34 @@ static void ex_folddo(exarg_T *eap)
 static void ex_terminal(exarg_T *eap)
 {
   char *name = (char *)p_sh;  // Default to 'shell' if {cmd} is not given.
+  char ex_cmd[1024] = { 0 };
```
We don't need to initialize `ex_cmd`. Both `if` branches start filling the array from position `0`.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9625,20 +9625,34 @@ static void ex_folddo(exarg_T *eap)
+  } else {
+    char **argv = build_argv((char *)p_sh);
+    char **p = argv;
+    char tempstring[512] = { 0 };
```
Does not need to be initialized.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -70,6 +73,25 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
   return rv;
 }
 
+/// Builds the argument vector for any command with with arguments.
```
Only one `with` needed.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -70,6 +73,25 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
+/// @return A newly allocated argument vector. It must also be freed with
+///         `shell_free_argv` when no longer needed.
+char **build_argv(char *cmd)
+  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC
```
Missing `FUNC_ATTR_NONNULL_ALL`.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -70,6 +73,25 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
+  char **rv = xmalloc((argc + 4) * sizeof(*rv));
+
+  tokenize((char_u *)cmd, rv);
+  assert(rv[0]);
```
If `cmd` is an empty string no pointer in `rv` is set. Looks like undefined behaviour, if we read `rv[0]`.

FriedSock replied:
According to f77f6449 it is not possible for shell, which is the only value that will be passed into cmd, to be null.

oni-link replied:
An empty string is not a `NULL` pointer, but has its first char set to `NUL`. For such a command string
`tokenize()` only returns `0` without writing into `rv`. Could happen with `:set sh=`.

FriedSock replied:
hmm, when I do  `:set shell=` then `:term`, I get `E474: Invalid argument` which I guess stems from a validation somewhere else. Do you think we should remove the assert?

oni-link replied:
The value of `rv[0]` is random at this point. You could use `xcalloc((argc + 4) * sizeof(*rv),1)` instead of `xmalloc()` here to set the memory to `0` and trigger the `assert()`. 

 I would remove the `assert()`, because the code never writes a `NULL` into `rv` up to this point.

FriedSock committed:
Fix comment typo and add missing macro

FriedSock committed:
Remove buildargv and handle when shell is unset

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -70,6 +73,24 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
+  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC FUNC_ATTR_NONNULL_ALL
+{
+  size_t argc = tokenize((char_u *)cmd, NULL);
+  char **rv = xmalloc((argc + 4) * sizeof(*rv));
```
`4` -> `1`. We only need a pointer for every argument and the terminating `NULL` pointer.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9624,26 +9624,34 @@ static void ex_folddo(exarg_T *eap)
+
+    snprintf(ex_cmd, sizeof(ex_cmd),
+             ":enew%s | call termopen([%s]) | startinsert",
+             eap->forceit ? "!" : "", shell_argv+1);
```
Missing spaces around `+`.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -9624,26 +9624,34 @@ static void ex_folddo(exarg_T *eap)
+             eap->forceit ? "!" : "", name);
+    xfree(name);
+  } else {
+    char **argv = build_argv((char *)p_sh);
```
Was there a specific reason why the new `build_argv(p_sh)` instead of `shell_build_argv(NULL,NULL)` was used?
```c
if (*p_sh == 0) {
   EMSG("No shell");
   return;
}

char **argv = shell_build_argv(NULL,NULL);
...
```

FriedSock replied:
Yes, I do believe I tried that.

I have tried adding it back now, without the early validation check and return, it looks like there is a hung build. Which might be because the assertion on `src/nvim/shell.c:71` would fail if shell was blank and cmd/extra args are both NULL.

Adding the early EMSG to avoid this seems to also cause test failures that could be due to early returning of `ex_terminal`. I'm going to revert back to using `build_argv`.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -41,6 +41,9 @@ typedef struct {
 /// Builds the argument vector for running the user-configured 'shell' (p_sh)
 /// with an optional command prefixed by 'shellcmdflag' (p_shcf).
 ///
+/// Creates ["shell", "-shellcommandflag", "-extra_args",
```
The order needs to be changed for `-shellcommandflag` and `-extra_args`.

justinmk commented:
@FriedSock The build isn't going to hang if an assertion fails. It means nvim is waiting for input, e.g. a "Press ENTER" prompt.

I can look into it (later) if you leave the PR as-is.

FriedSock commented:
@justinmk Ah right, I will also take a look at it later today and won't change it until there is a definite fix 👍

justinmk commented:
Merged. Thanks @FriedSock  and  @oni-link !
